### PR TITLE
Upgrade to Python3

### PR DESCRIPTION
**Pull request by kentsanggds**
## What

Upgraded to python 3 and migrated from nosetests to pytests. 

There may still be bits of pylons left in the code to tidy up, the focus of this PR is to migrate the relevant tests to pytest and get them working. 

There is in the meantime an issue with trying to get the branch to pass on the CI stack because its not set up for python 3, so will be fixing that soon.

The README has been updated with a section on how to do the testing with docker ckan.

NOTE - This probably won't be merged into `ckan-2.9` but will be used as a base for `master-2.9` so ignore the conflicts at the bottom of the PR.

## Reference 

https://trello.com/c/AVApiS3c/88-get-tests-for-datagovuk-extension-passing

### CODE REVIEW

**Review by oscarwyatt — CHANGES_REQUESTED** [review]
Some minor comments but otherwise great work 🏅 

I did a bit of research about the problem we discussed with beautifulsoup. The best solution I've found is to run `pip install --upgrade beautifulsoup4`, I'm not certain where that would be run though so I'll leave that to you to implement

This is probably beyond my remit so apologies if it's a tad presumptuous but as a general point (and certainly not something that you should do as part of this PR) I think it'd be a good thing to add a linter at some point. I'll probably cause a fair bit of extra work to fix the things it spots but outside of your commits in this PR I've spotted a few things that seem to be a bit widespread in the codebase that a linter would pick up on and that would be good to address (mostly minor things, for example redundant imports or variables that aren't used - I expect there will be plenty of other things though)

File: ckanext/datagovuk/nosetests/test_package.py [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        app = helpers._get_test_app()
+        env = {'REMOTE_USER': user['name'].encode('ascii')}
+        response = app.get(
+            url_for(controller='package',
+                    action='new_resource',
+                    id=dataset['name']),
+            extra_environ=env,
+        )
+
+        page = BeautifulSoup(response.html.decode('utf-8'), 'html.parser')
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply ckanext/datagovuk/nosetests/test_package.py:114]: nosetests will be dropped in another PR, they are just here for reference to compare with pytest

File: ckanext/datagovuk/nosetests/test_package.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply ckanext/datagovuk/nosetests/test_package.py:87]: nosetests will be dropped in another PR, they are just here for reference to compare with pytest

File: ckanext/datagovuk/nosetests/test_package.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
+from copy import deepcopy
+import json
+
+from bs4 import BeautifulSoup
```
[the remark this answers is not in the document]
Comment (kentsanggds) [reply ckanext/datagovuk/nosetests/test_package.py:4]: nosetests will be dropped in another PR, they are just here for reference to compare with pytest

**Comment by kentsanggds** [comment]
I take onboard your suggestion of adding a linter, but as I've got a tight deadline to finish the upgrade work, I'll have to do that if I have time after completing the upgrade.

**Review by oscarwyatt — APPROVED** [review]
Great stuff 👍

**Comment by kentsanggds** [comment]
Commits here have been pushed to a new branch - `master-2.9`, will close this PR as approved

**Pull request closed without merging**